README.md
<p align="center">
  <img src="https://github.com/erebe/greenclip/raw/master/logo.png" alt="logo"/>
</p>

## Description

Recycle your clipboard selections with greenclip and don't waste your time anymore
to reselect things over and over.

**Purpose:**
Keeps track of your history of selections to quickly switch between them

**Demo:**  <a href="https://www.youtube.com/watch?v=Utk-9Gy8H3w">Video Link</a>
<br/>
<a href="https://www.youtube.com/watch?v=4IycORAdW9M">Video made by @gotbletu</a>

**Features:**
  + Integrated with [rofi](https://github.com/DaveDavenport/rofi)
  + Permanently set some selections to added at the end (set `static_history_file = your/file/with/static/entries` in the config file)
  + Merge X Primary selection with clipboard selection (set `use_primary_selection_as_input = true` in the config file)
  + Blacklist some applications (see `I want to blacklist some applications !` in the FAQ section)
  + Copy small images (you can disable it in the config)

## Installation

- It's a static binary so drop it anywhere in your $PATH env

```wget https://github.com/erebe/greenclip/releases/download/4.1/greenclip```


- Alternatively if you are using Archlinux you can install the package from AUR

``yay rofi-greenclip``

PS: If you want, you can add a permanent list of selections to be added to your current history. Go see the config file

Configuration file can be found at: 
```
❯ cat ~/.config/greenclip.toml

[greenclip]
  history_file = "/home/erebe/.cache/greenclip.history"
  max_history_length = 50
  max_selection_size_bytes = 0
  trim_space_from_selection = true
  use_primary_selection_as_input = false
  blacklisted_applications = []
  enable_image_support = true
  # path without ending / will generate mktemp directory at the prefix location
  image_cache_directory = "/tmp/greenclip"
  static_history = [
 '''¯\_(ツ)_/¯''',
]
```

## Usage

Greenclip is intended to be used with [rofi](https://github.com/DaveDavenport/rofi)

1. Spawn the daemon ``` greenclip daemon ```
2. When ever you need to get your selections history ``` rofi -modi "clipboard:greenclip print" -show clipboard -run-command '{cmd}' ```
3. The entry that you have selected will be in your clipboard now
4. Configuration file can be found in ```.config/greenclip.cfg```

## Migrating from 2.x version to 3.x one

1. Kill all greenclip process already running
2. Delete your old config file and old cache file (in ~/.cache/greenclip*)
3. Start the new version of greenclip

## Building the project

1. Get [stack](https://docs.haskellstack.org/en/stable/README/) for Haskell
2. stack init && stack install

or you can look at the .travis.yml file

## FAQ

Q. **Greenclip does not copy selection !**

A. Greenclip cannot run alongside other clipboard manager. If you have already one activated (there is one by default in KDE), you have first to disable it before trying to run greenclip

---------

Q. **I cannot paste images !**

A. You can only keep in your history small images  (~ <500 kb). I disabled the support for bigger images as it will turn your CPU into a heater for winter.

----------

Q. **I want to blacklist some applications !**

A. There is in your config file a section for blacklisting applications.
   You can run greenclip daemon in a terminal to find what is the name of the application who has the current selection.
   Be aware that a lot of application does not name their clipboard process/window, so it will be empty most of the time.
   Be aware also, that if you use an electron application (like slack i.e) you will get a generic name like "Chromium clipboard"

----------

Q. **I want to paste selection directly after selecting them !**

A. This is not possible as when you invoke rofi, you lose focus of your current window and there is no way to find it back (from greenclip point of view)

   Nonetheless, you can emulate the feature with xdotool `rofi -modi "clipboard:greenclip print" -show clipboard -run-command '{cmd}' ; sleep 0.5; xdotool type $(xclip -o -selection clipboard)`

   or look at this issue for a more complete solution https://github.com/erebe/greenclip/issues/27

----------

Q. **I got a strange behavior when clearing the clipboard - greenclip clear**

A. The command is not doing what it is supposed to do, if greenclip's daemon is running. You need to stop it first.

   Use this command to clear the history `pkill greenclip && greenclip clear && greenclip daemon &`
   
   For more information regarding why, refer to https://github.com/erebe/greenclip/issues/34 

----------

Q. **I need multi-line text in staticHistory**

A. https://github.com/erebe/greenclip/issues/78

----------

Q. **I don't want to use rofi !**

A. You can also use greenclip with [dmenu](http://tools.suckless.org/dmenu) or [fzf](https://github.com/junegunn/fzf). Example usage:

   `greenclip print | sed '/^$/d' | dmenu -i -l 10 -p clipboard | xargs -r -d'\n' -I '{}' greenclip print '{}'`

   `greenclip print | sed '/^$/d' | fzf -e | xargs -r -d'\n' -I '{}' greenclip print '{}'`

----------

Q. **I want to thank you for greenclip !**

A. Give a star to this repository


changelog
4.0
 * Switch config file format to toml
 * Add an option to limit the size of the clipboard entries
 * Inline static history inside the config file
 * Do not override config file at startup

3.2
  * Add option to not trim space from selection

3.1
  * Chmod to 600 the cache history file
  * Do not rely on atom given by thirdapp to avoid buggy app (thermite)

3.0
  * Support for blacklisting applications
  * Support for keeping images in history
  * Fix exception with clear commands

2.1
  * Support ~ expansion in config file
  
2.0
  * X Primary selection can be used as Input

1.3
  * Fix config file parsing
  * Add version in the help message

1.2
  * Avoid that the program hang sometime due to the fact that XnextEvent is uninterruptible

1.1
  * Better Unicode support
  * Add timeout on how long we can wait for selection
  * Correctly abort the program when no XDisplay is available
  * Avoid re-opening a connection to the X server each time (dirty-fix)
  * Avoid hanging issue

1.0
  * Working version


src/Config.hs
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NoImplicitPrelude     #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE GADTs                 #-}
{-# LANGUAGE Rank2Types            #-}
{-# LANGUAGE NoStrict              #-}
{-# LANGUAGE OverloadedStrings     #-}


module Config (Config(..), Pos(..), parseConfig) where

import           Protolude                        hiding (readFile, to, (<&>), (&))

import qualified Data.ByteString.Char8            as BS
import qualified Data.Text                        as T
import           Data.Char                        (ord, isDigit, isPrint)
import           Lens.Micro
import           Control.Applicative.Combinators
import           Control.Applicative.Permutations

-- The Configuration ----------------------------------------------------------

data Config = Config
  { maxHistoryLength           :: Int
  , historyPath                :: Text
  , staticHistoryPath          :: Text
  , imageCachePath             :: Text
  , usePrimarySelectionAsInput :: Bool
  , blacklistedApps            :: [Text]
  , trimSpaceFromSelection     :: Bool
  } deriving (Show, Read)

data Pos = Pos
         { line      :: Integer -- The line of the character
         , character :: Integer -- The position in the line of the character
         , offset    :: Integer -- The position in the global flux
         } deriving (Show)

parseConfig :: ByteString -> Either Pos Config
parseConfig file = case runParser file configP of
                     Left pos   -> Left pos
                     Right endo -> Right $ appEndo endo initConfig
  where initConfig :: Config
        initConfig = Config 0 "" "" "" True [] True

-- Parser configuration -------------------------------------------------------

maxHistoryLengthLens :: Lens' Config Int
maxHistoryLengthLens = lens maxHistoryLength $ \pc i -> pc { maxHistoryLength = i }

historyPathLens :: Lens' Config Text
historyPathLens = lens historyPath $ \pc t -> pc { historyPath = t }

staticHistoryPathLens :: Lens' Config Text
staticHistoryPathLens = lens staticHistoryPath $ \pc t -> pc { staticHistoryPath = t }

imageCachePathLens :: Lens' Config Text
imageCachePathLens = lens imageCachePath $ \pc t -> pc { imageCachePath = t }

usePrimarySelectionAsInputLens :: Lens' Config Bool
usePrimarySelectionAsInputLens = lens usePrimarySelectionAsInput
                                    $ \pc b -> pc { usePrimarySelectionAsInput = b }

blacklistedAppsLens :: Lens' Config [Text]
blacklistedAppsLens = lens blacklistedApps $ \pc l -> pc { blacklistedApps = l }

trimSpaceFromSelectionLens :: Lens' Config Bool
trimSpaceFromSelectionLens = lens trimSpaceFromSelection $ \pc b -> pc { trimSpaceFromSelection = b }

data ConfigMemberParser where
    PCMP :: Parser a       -- How to parse the config option
         -> Lens' Config a -- How to set it
         -> Text           -- The default name for this config option
         -> [ Text ]       -- Aliases for the config option
         -> Maybe a        -- An eventual default value
         -> ConfigMemberParser

configOptions :: [ConfigMemberParser]
configOptions = [ PCMP intP            maxHistoryLengthLens           "maxHistoryLength" 
                       [ ]
                       $ Just 50
                , PCMP stringP         historyPathLens                "historyPath" 
                       [ "historyFile" ]
                       $ Just "~/.cache/greenclip.history"
                , PCMP stringP         staticHistoryPathLens          "staticHistoryPath"
                       [ "staticHistoryFile" ]
                       $ Just "~/.cache/greenclip.staticHistory"
                , PCMP stringP         imageCachePathLens             "imageCachePath"
                       [ "imageCacheDirectory" ]
                       $ Just "/tmp/greenclip/"
                , PCMP boolP           usePrimarySelectionAsInputLens "usePrimarySelectionAsInput"
                       [ ]
                       $ Just False
                , PCMP (listP stringP) blacklistedAppsLens            "blacklistedApps"
                       [ ]
                       $ Just [ ]
                , PCMP boolP           trimSpaceFromSelectionLens     "trimSpaceFromSelection"
                       [ ]
                       $ Just True
                ]

-- The Parser -----------------------------------------------------------------

isNewLine :: Char -> Bool
isNewLine w = w == '\n'

incrementPos :: Pos -> Char -> Pos
incrementPos (Pos ln ch off) c = if isNewLine c
                                    then Pos (ln + 1) 0        (off + 1)
                                    else Pos ln       (ch + 1) (off + 1)

-- An input is a bytestring with a position representing the position
-- of the first character in the complete input
type Input = (ByteString, Pos)

newtype Parser a = Parser (Input -> (Either Pos a,Input))

tryP :: Parser a -> Parser a
tryP (Parser f) = Parser $ \input -> 
    case f input of
      (Left p, _) -> (Left p, input)
      result      -> result

getP :: Parser Char
getP = Parser $ \(text,pos) -> if BS.null text
                                  then (Left pos, (text,pos))
                                  else let c = BS.head text
                                       in (Right c, (BS.tail text, incrementPos pos c))

eofP :: Parser ()
eofP = Parser $ \(input,pos) ->
    if BS.null input then (Right (), (input,pos)) else (Left pos, (input,pos))

runParser :: ByteString -> Parser a -> Either Pos a
runParser text (Parser f) = fst $ f (text,pos)
    where pos :: Pos
          pos = Pos 1 0 0

instance Functor Parser where
    fmap f (Parser g) = Parser $ \input -> case g input of
                                             (Right x, out) -> (Right (f x), out)
                                             (Left e,  out) -> (Left e,      out)

instance Applicative Parser where
    (Parser f) <*> x = Parser $ \input ->
        let (result, out) = f input in
        case result of
          Right fn -> let Parser g = fn <$> x in g out
          Left e   -> (Left e, out)
    pure x = Parser $ \input -> (Right x, input)

instance Alternative Parser where
    empty = Parser $ \(text,pos) -> (Left pos, (text,pos))
    (Parser f) <|> (Parser g) = Parser $ \input ->
        case f input of
          (Left _, ninput) -> g ninput
          result           -> result

instance Monad Parser where
    (Parser f) >>= g = Parser $ \input ->
        case f input of
          (Left e, out)  -> (Left e, out)
          (Right x, out) -> let Parser gn = g x in gn out

instance MonadPlus Parser where

-- Config Parser --------------------------------------------------------------

condP :: (Char -> Bool) -> Parser Char
condP cond = getP >>= \r -> guard (cond r) >> return r

digitP :: Parser Int
digitP = condP isDigit >>= \c -> return (ord c - ord '0')

intP :: Parser Int
intP = do
    numbers <- reverse <$> some (tryP digitP)
    let (num,_) = foldl (\(n,o) d -> (n + d*o, o*10)) (0,1) numbers
    return num

whiteP :: Parser ()
whiteP = void $ condP $ \c -> c == ' ' || c == '\t' || c == '\n' || c == '\r'

whitesP :: Parser ()
whitesP = skipMany $ tryP whiteP

charP :: Char -> Parser ()
charP c = void $ condP (==c)

textP :: Text -> Parser ()
textP = mapM_ charP . T.unpack

stringP :: Parser Text
stringP = between open close $ T.pack <$> alphasP
    where open :: Parser ()
          open = whitesP >> charP '"'
          close :: Parser ()
          close = charP '"' >> whitesP
          alphasP :: Parser [Char]
          alphasP = many $ tryP $ condP $ \c -> isPrint c && (c /= '"') 

commaP :: Parser ()
commaP = whitesP >> tryP (charP ',') >> whitesP

listP :: Parser a -> Parser [a]
listP pel = between (charP '[' >> whitesP) (whitesP >> charP ']')
                  $ sepBy (tryP pel) commaP

boolP :: Parser Bool
boolP = ( tryP (textP "True") >> return True  )
    <|> ( textP "False"       >> return False )

pcmpP :: ConfigMemberParser -> Parser (Endo Config)
pcmpP (PCMP parser lense name aliases _) = do
    foldl (\p alias -> p <|> tryP (textP alias)) (tryP $ textP name) aliases
    whitesP
    charP '='
    whitesP
    r <- parser
    return $ Endo $ \pc -> pc & lense .~ r

pcmpPerm :: ConfigMemberParser -> Permutation Parser (Endo Config)
pcmpPerm pcmp@(PCMP _ lense _ _ mdef) =
    case mdef of
      Nothing  -> toPermutation $ tryP $ pcmpP pcmp
      Just def -> toPermutationWithDefault (mkDef def lense) $ tryP $ pcmpP pcmp
  where mkDef :: a -> Lens' Config a -> Endo Config
        mkDef df ls = Endo $ \pc -> pc & ls .~ df

optionsP :: [ConfigMemberParser] -> Permutation Parser (Endo Config)
optionsP (hopt : topts) = foldl (\perm pcmp -> (<>) <$> pcmpPerm pcmp <*> perm)
                                (pcmpPerm hopt) topts
optionsP [] = toPermutation empty -- Won't happen

configP :: Parser (Endo Config)
configP = do
    textP "Config"
    whitesP
    result <- between (charP '{' >> whitesP) (whitesP >> charP '}')
                    $ intercalateEffect commaP $ optionsP configOptions
    whitesP
    eofP
    return result



lib/Clipboard/System/Clipboard/X11.hs

{-# LANGUAGE RecordWildCards #-}

module System.Clipboard.X11
  ( getClipboardString
  , setClipboardString
  , getPrimaryClipboard
  ) where

import Graphics.X11.Xlib
import Graphics.X11.Xlib.Extras
import System.Posix.Process     (forkProcess)

import Codec.Binary.UTF8.String (decode, encode)
import Control.Monad
import Data.Maybe
import Foreign                  (peekByteOff)
import Foreign.C.Types          (CChar, CUChar)
import Foreign.Marshal.Array    (withArrayLen)
import System.Directory         (setCurrentDirectory)
import System.IO                (hClose, stderr, stdin, stdout)
import System.IO.Unsafe (unsafePerformIO)
import Data.IORef
import Control.Concurrent (threadDelay)

getClipboardString :: IO (Maybe String)
getClipboardString = do
    (display, window, clipboards) <- readIORef initialSetup
    inp <- internAtom display "clipboard_get" False
    target <- internAtom display "UTF8_STRING" True
    xConvertSelection display (head clipboards) target inp window currentTime
    Just <$> clipboardInputWait display window inp

getPrimaryClipboard :: IO (Maybe String)
getPrimaryClipboard = do
    (display, window, clipboards) <- readIORef initialSetup
    inp <- internAtom display "clipboard_get" False
    target <- internAtom display "UTF8_STRING" True
    xConvertSelection display (clipboards !! 1) target inp window currentTime
    Just <$> clipboardInputWait display window inp

clipboardInputWait :: Display -> Window -> Atom -> IO String
clipboardInputWait display' window' inp' = allocaXEvent (go display' window' inp')
  where
  go display window inp evPtr = do
    waitForEvent display
    nextEvent display evPtr
    ev <- getEvent evPtr
    if ev_event_type ev == selectionNotify
       then charsToString . fromMaybe mempty <$> getWindowProperty8 display inp window
       else go display window inp evPtr
  waitForEvent display = do
    nbEvs <- pending display
    when (nbEvs == 0) $ threadDelay 100000 >> waitForEvent display

charsToString :: [CChar] -> String
charsToString = decode . map fromIntegral

setClipboardString :: String -> IO ()
setClipboardString str = void $ forkProcess $ do
        mapM_ hClose [stdin, stdout, stderr]
        setCurrentDirectory "/"
        (display, window, clipboards) <- readIORef initialSetup
        mapM_ (\atom -> xSetSelectionOwner display atom window currentTime) clipboards
        advertiseSelection display clipboards (stringToChars str)
        cleanup display window

advertiseSelection :: Display -> [Atom] -> [CUChar] -> IO ()
advertiseSelection display clipboards' str = allocaXEvent (go clipboards')
  where
    go [] _ = return ()
    go clipboards evPtr = do
      nextEvent display evPtr
      ev <- getEvent evPtr
      case ev of
          SelectionRequest {..} -> do
              target' <- getAtomName display ev_target
              res <- handleOutput display ev_requestor ev_property target' str
              sendSelectionNotify display ev_requestor ev_selection ev_target res ev_time
              go clipboards evPtr

          _ | ev_event_type ev == selectionClear -> do
              target <- peekByteOff evPtr 40 :: IO Atom
              go (filter (/= target) clipboards) evPtr

          _ -> go clipboards evPtr

handleOutput :: Display -> Window -> Atom -> Maybe String -> [CUChar] -> IO Atom
handleOutput display req prop (Just "UTF8_STRING") str = do
    prop' <- getAtomName display prop
    if isNothing prop' then handleOutput display req prop Nothing str else do
        target <- internAtom display "UTF8_STRING" True
        void $ withArrayLen str $ \len str' ->
            xChangeProperty display req prop target 8 propModeReplace str'
                            (fromIntegral len)
        return prop
handleOutput _ _ _ _ _ = return none

sendSelectionNotify :: Display -> Window -> Atom -> Atom -> Atom -> Time -> IO ()
sendSelectionNotify display req sel target prop time = allocaXEvent $ \ev -> do
    setEventType ev selectionNotify
    setSelectionNotify ev req sel target prop time
    sendEvent display req False 0 ev

stringToChars :: String -> [CUChar]
stringToChars = map fromIntegral . encode

initialSetup :: IORef (Display, Window, [Atom])
initialSetup = unsafePerformIO $ do
    display <- openDisplay ""
    window <- createSimpleWindow display (defaultRootWindow display)
                                 0 0 1 1 0 0 0
    clipboards <- internAtom display "CLIPBOARD" False
    newIORef (display, window, [clipboards, pRIMARY])
{-# NOINLINE initialSetup #-}

cleanup :: Display -> Window -> IO ()
cleanup display window = do
    destroyWindow display window
    closeDisplay display


lib/Clipboard/System/Clipboard/Windows.hs
-- | System clipboard interface with unicode support.
--
-- For more information, see "Graphics.Win32.GDI.Clip"
-- or documentation for /GetClipboardData/ on MSDN.
module System.Clipboard.Windows
    ( setClipboardString
    , getClipboardString
    ) where

import System.Win32.Mem
    (globalAlloc, globalLock, globalUnlock, copyMemory, gHND)
import Graphics.Win32.GDI.Clip
    ( openClipboard, closeClipboard, emptyClipboard,
      getClipboardData, setClipboardData
    , cF_TEXT, ClipboardFormat
    , isClipboardFormatAvailable)
import Foreign.C
   (withCAString, peekCAString, withCWString, peekCWString)
import Foreign.Ptr
    (castPtr, nullPtr)
import Data.List
    (genericLength)
import Control.Exception
    (bracket_, bracket)
import Data.Maybe
    (isJust)

cF_UNICODETEXT :: ClipboardFormat
cF_UNICODETEXT = 13

setClipboardString :: String -> IO ()
setClipboardString str =
   withCWString str $ \cstring -> do
   mem <- globalAlloc gHND strLen
   bracket (globalLock mem) globalUnlock $ \mem' -> do
       copyMemory mem' (castPtr cstring) strLen
       bracket_ (openClipboard nullPtr) closeClipboard $ do
           emptyClipboard
           setClipboardData cF_UNICODETEXT mem'
           return ()
 where
   strLen = 2 * (genericLength str + 1)

getClipboardString :: IO (Maybe String)
getClipboardString =
   bracket_ (openClipboard nullPtr) closeClipboard $ do
     isUnicodeAvailable <- isClipboardFormatAvailable cF_UNICODETEXT
     if isUnicodeAvailable
       then do handle <- getClipboardData cF_UNICODETEXT
               mem <- globalLock handle
               str <- peekCWString (castPtr mem)
               globalUnlock mem
               return $ Just str
       else do isAnsiAvailable <- isClipboardFormatAvailable cF_TEXT
               if isAnsiAvailable
                 then do handle <- getClipboardData cF_TEXT
                         mem <- globalLock handle
                         str <- peekCAString (castPtr mem)
                         globalUnlock mem
                         return $ Just str
                 else return Nothing


src/Main.hs

{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE FlexibleContexts      #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE NoImplicitPrelude     #-}
{-# LANGUAGE OverloadedStrings     #-}
{-# LANGUAGE ScopedTypeVariables   #-}
{-# LANGUAGE ViewPatterns          #-}


module Main where

import           Protolude             hiding (readFile, to, (<&>), (&), toS)
import           Protolude.Conv        (toS)

import           Control.Monad.Catch   (MonadCatch, catchAll)
import           Data.Binary           (decodeFile, encode)
import qualified Data.ByteString       as B
import           Data.List             (dropWhileEnd)
import qualified Data.Text             as T
import           Data.Vector           (Vector)
import qualified Data.Vector           as V
import           Lens.Micro
import           Lens.Micro.Mtl    hiding ((.=))
import qualified System.Directory      as Dir
import           System.Posix.Files    (setFileMode)
import           System.Environment    (lookupEnv)
import           System.IO             (hClose, hGetContents)
import           System.Timeout        (timeout)
import           System.Wordexp.Simple (wordexp)
import           System.Posix.Temp     (mkdtemp)

import Toml (TomlCodec, (.=))
import qualified Toml

import qualified Clipboard             as Clip


data Command = DAEMON | PRINT | COPY Text | CLEAR | HELP deriving (Show, Read)

data Config = Config
  { maxHistoryLength           :: Int
  , maxItemSizeBytes           :: Int
  , historyPath                :: Text
  , imageCachePath             :: Text
  , usePrimarySelectionAsInput :: Bool
  , blacklistedApps            :: [Text]
  , trimSpaceFromSelection     :: Bool
  , enableImageSupport         :: Bool
  , staticHistory              :: [Text]
  } deriving (Show, Read)

configCodec :: TomlCodec Config
configCodec = Config
    <$> Toml.int "max_history_length"  .= maxHistoryLength
    <*> Toml.int "max_selection_size_bytes" .= maxItemSizeBytes
    <*> Toml.text "history_file" .= historyPath
    <*> Toml.text "image_cache_directory" .= imageCachePath
    <*> Toml.bool "use_primary_selection_as_input" .= usePrimarySelectionAsInput
    <*> Toml.arrayOf Toml._Text  "blacklisted_applications" .= blacklistedApps
    <*> Toml.bool "trim_space_from_selection" .= trimSpaceFromSelection
    <*> Toml.bool "enable_image_support" .= enableImageSupport
    <*> Toml.arrayOf Toml._Text  "static_history" .= staticHistory


type ClipHistory = Vector Clip.Selection

readFile :: FilePath -> IO ByteString
readFile filepath = bracket (openFile filepath ReadMode) hClose $ \h -> do
  str <- hGetContents h
  return $! toS str


getHistory :: (MonadIO m, MonadReader Config m) => m ClipHistory
getHistory = do
  storePath <- view $ to (toS . historyPath)
  liftIO $ (V.fromList <$> decodeFile storePath) `catchAll` const mempty


getStaticHistory :: (MonadIO m, MonadReader Config m) => m ClipHistory
getStaticHistory = do
  history <- view $ to (staticHistory)
  return . V.fromList $ Clip.Selection "greenclip" . Clip.UTF8 <$> history



storeHistory :: (MonadIO m, MonadReader Config m) => ClipHistory -> m ()
storeHistory history = do
  storePath <- view $ to (toS . historyPath)
  liftIO $ writeH storePath history
  where
    writeH storePath = B.writeFile storePath . toS . encode . V.toList


appendToHistory :: (MonadIO m, MonadReader Config m) => Clip.Selection -> ClipHistory -> m (ClipHistory, ClipHistory)
appendToHistory sel history' = do
  trimSelection <- view $ to trimSpaceFromSelection
  case sel of
    Clip.Selection appName (Clip.UTF8 txt) -> appendGeneric (if trimSelection then Clip.Selection appName (Clip.UTF8 (T.strip txt)) else sel) history'
    Clip.Selection _ (Clip.PNG bytes) -> appendImage Clip.PNG ".png" bytes
    Clip.Selection _ (Clip.JPEG bytes) -> appendImage Clip.JPEG ".jpeg" bytes
    Clip.Selection _ (Clip.BITMAP bytes) -> appendImage Clip.BITMAP ".bmp" bytes


  where
    appendImage imgCtr extension bytes = do
      cachePth <- view (to imageCachePath)
      let imgHash = show $ hash bytes
      let imgPath = toS $ cachePth <> "/" <> imgHash <> extension
      _ <- liftIO $ writeImage imgPath bytes
      appendGeneric (sel {Clip.selection = imgCtr $ toS imgHash}) history'

    writeImage path bytes = do
      fileExist <- Dir.doesFileExist path
      if fileExist
        then return False
        else B.writeFile path bytes >> setFileMode path 0o600 >> return True

    appendGeneric selection history =
      if maybe False (\sel' -> Clip.selection sel' == Clip.selection selection) (history V.!? 0)
        then return (history, mempty)
        else do
          maxLen <- view $ to maxHistoryLength
          return $ V.splitAt maxLen . V.cons selection $ V.filter (\ori -> Clip.selection ori /= Clip.selection selection) history


setHistoryFilePermission :: (MonadIO m, MonadReader Config m) => m ()
setHistoryFilePermission = do
  storePath <- view $ to (toS . historyPath)
  fileExist <- liftIO $ Dir.doesFileExist storePath
  when (not fileExist) (storeHistory mempty)
  liftIO $ setFileMode storePath 0o600

prepareDirs :: (MonadIO m, MonadReader Config m) => m ()
prepareDirs = do
  historyFile <- view $ to (T.unpack . historyPath)
  imgDir <- view $ to (T.unpack . imageCachePath)
  let dirs = [imgDir, dropWhileEnd (/= '/') historyFile]
  mapM_ (liftIO . Dir.createDirectoryIfMissing True) dirs

runDaemon:: (MonadIO m, MonadCatch m, MonadReader Config m) => m ()
runDaemon = prepareDirs >> setHistoryFilePermission >> (forever $ go `catchAll` handleError)
  where
    _0_5sec :: Int
    _0_5sec = 5 * 100000

    _5sec :: Int
    _5sec = 5000000

    go = do
      history <- getHistory
      usePrimary <- view $ to usePrimarySelectionAsInput
      enableImage <- view $ to enableImageSupport
      cfg <- ask

      liftIO $ bracket Clip.getXorgContext Clip.destroyXorgContext $ \x11Context -> do
        let getSelections = (getSelectionFrom (Clip.getClipboardSelection x11Context enableImage), Nothing)
                          : [(getSelectionFrom (Clip.getPrimarySelection x11Context enableImage), Nothing) | usePrimary]
        void $ runReaderT (innerloop getSelections history) cfg

    getSelection [] = return ([], Nothing)
    getSelection ((getSel, lastSel):getSels) = do
      selection <- liftIO getSel
      if fmap Clip.selection selection /= fmap Clip.selection lastSel
         then return ((getSel, selection) : getSels, selection)
         else getSelection getSels >>= \(e, sel) -> return ((getSel, lastSel) : e, sel)

    innerloop :: (MonadIO m, MonadReader Config m) => [(IO (Maybe Clip.Selection), Maybe Clip.Selection)] -> ClipHistory -> m ClipHistory
    innerloop getSelections history = do
      -- Get selection from enabled clipboards
      (getSelections', rawSelection) <- liftIO $ getSelection getSelections

      -- Do not store selection items above threshold size
      maxItemSize <- view (to maxItemSizeBytes)
      let sel = case rawSelection of
            Nothing -> Nothing
            Just selection -> if maxItemSize > 0 && Clip.selectionLength selection >= maxItemSize
                              then Nothing
                              else Just selection
             
      -- Do not use selection coming from blacklisted app
      liftIO $ when (isJust sel) (print (Clip.appName <$> sel))
      blacklist <- view (to blacklistedApps)
      let selection = sel >>= \s -> if isJust $ find (== Clip.appName s) blacklist
                                      then Nothing
                                      else Just s

      -- Append current selection to history and get back entries needed to be purged
      (history', toBePurged) <- maybe (return (history, mempty)) (`appendToHistory` history) selection
      traverse_ purgeSelection toBePurged

      -- backup on disk history if it as changed seen last backup
      when (isJust selection && history' /= history) (storeHistory history')

      -- Getting some rest
      liftIO $ threadDelay _0_5sec
      innerloop getSelections' history'

    getSelectionFrom :: IO (Maybe Clip.Selection) -> IO (Maybe Clip.Selection)
    getSelectionFrom = fmap join . timeout _5sec

    purgeSelection (Clip.Selection _ (Clip.PNG txt)) = purge (toS txt <> ".png")
    purgeSelection (Clip.Selection _ (Clip.JPEG txt)) = purge (toS txt <> ".jpeg")
    purgeSelection (Clip.Selection _ (Clip.BITMAP txt)) = purge (toS txt <> ".bmp")
    purgeSelection _ = return ()

    purge path = do
      cachePth <- view (to imageCachePath)
      liftIO $ Dir.removeFile (toS $ cachePth <> "/" <> path) `catchAll` const mempty

    handleError ex = do
      let displayMissing = "openDisplay" `T.isInfixOf` show ex
      if displayMissing
      then panic "X display not available. Please start Xorg before running greenclip"
      else print ex
      liftIO $ threadDelay _0_5sec


toRofiStr :: Clip.Selection -> Text
toRofiStr (Clip.Selection _ (Clip.UTF8 txt)) = T.map (\c -> if c == '\n' || c == '\r' then '\xA0' else c) txt
toRofiStr (Clip.Selection appName (Clip.PNG txt)) = "image/png " <> appName <> " " <> toS txt
toRofiStr (Clip.Selection appName (Clip.JPEG txt)) = "image/jpeg " <> appName <> " " <> toS txt
toRofiStr (Clip.Selection appName (Clip.BITMAP txt)) = "image/bmp " <> appName <> " " <> toS txt

fromRofiStr :: Text -> Text -> IO Clip.Selection
fromRofiStr cachePth txt@(T.isPrefixOf "image/png " -> True) = B.readFile (toS $ cachePth <> "/" <> getHash txt <> ".png") <&> Clip.Selection "greenclip" . Clip.PNG
fromRofiStr cachePth txt@(T.isPrefixOf "image/jpeg " -> True) = B.readFile (toS $ cachePth <> "/" <> getHash txt <> ".jpeg") <&> Clip.Selection "greenclip" . Clip.JPEG
fromRofiStr cachePth txt@(T.isPrefixOf "image/bmp " -> True) = B.readFile (toS $ cachePth <> "/" <> getHash txt <> ".bmp") <&> Clip.Selection "greenclip" . Clip.BITMAP
fromRofiStr _ txt = return $ Clip.Selection "greenclip" (Clip.UTF8 (T.map (\c -> if c == '\xA0' then '\n' else c) txt))

getHash :: Text -> Text
getHash = fromMaybe mempty . lastMay . T.split (== ' ')


printHistoryForRofi :: (MonadIO m, MonadReader Config m) => m ()
printHistoryForRofi = do
  history <- mappend <$> getHistory <*> getStaticHistory
  _ <- traverse (putStrLn . toRofiStr) history
  return ()


advertiseSelection :: (MonadIO m, MonadReader Config m) => Text -> m ()
advertiseSelection txt = do
  cachePth <- view (to imageCachePath)
  selection <- liftIO $ fromRofiStr cachePth txt
  liftIO $ Clip.setClipboardSelection selection


getConfig :: IO Config
getConfig = do
  home <- Dir.getHomeDirectory
  let configPath = home <> "/.config/greenclip.toml"
  
  configExist <- Dir.doesFileExist configPath
  when (not configExist) $ do 
    config <- Toml.encode (Toml.table configCodec "greenclip") <$> defaultConfig
    writeFile configPath config
    return ()
  
  tomlRes <- Toml.decodeFileEither (Toml.table configCodec "greenclip") configPath
  when (isLeft tomlRes) $ do
    die . toS $  "Error parsing the config file at " <> (show configPath) <> "\n" <> Toml.prettyTomlDecodeErrors (fromLeft mempty tomlRes)
  
  let cfg = fromRight (Config 50 0 "" "" False [] True True []) tomlRes 
  
  -- Replace $HOME|~|*... in config path
  cfg <- do
    imgCachePath <- wordexp . toS $ imageCachePath cfg
    historyP <- wordexp . toS $ historyPath cfg 
    return $ cfg { imageCachePath = (toS $ headDef "" imgCachePath), historyPath = (toS $ headDef "" historyP)}
    
  -- if it ends with / we don't create a temp directory
  -- user is responsible for it
  cfg <- if (lastDef ' ' (toS $ imageCachePath cfg) /= '/') 
      then do
        dirPath <- mkdtemp $ (toS $ imageCachePath cfg)
        return $ cfg { imageCachePath = toS dirPath }
      else return cfg
     
  return cfg

  where
    defaultConfig = do 
      homeDir <- toS . fromMaybe mempty . listToMaybe <$> wordexp "~/"
      return $ Config 50 0 (homeDir <> ".cache/greenclip.history") "/tmp/greenclip" False [] True True 
        ["Greenclip has been updated to v4.1, update your new config file at ~/.config/greenclip.toml"]


parseArgs :: [Text] -> Command
parseArgs ("daemon":_)   = DAEMON
parseArgs ["clear"]      = CLEAR
parseArgs ["print"]      = PRINT
parseArgs ["print", sel] = COPY sel
parseArgs _              = HELP

run :: Command -> IO ()
run cmd = do
  cfg <- getConfig
  case cmd of
    DAEMON   -> runReaderT runDaemon cfg
    PRINT    -> runReaderT printHistoryForRofi cfg
    CLEAR    -> runReaderT (storeHistory mempty) cfg
    -- Should rename COPY into ADVERTISE but as greenclip is already used I don't want to break configs
    -- of other people
    COPY sel -> runReaderT (advertiseSelection sel) cfg
    HELP     -> putText $ "greenclip v4.1 -- Recyle your clipboard selections\n\n" <>
                          "Available commands\n" <>
                          "daemon: Spawn the daemon that will listen to selections\n" <>
                          "print:  Display all selections history\n" <>
                          "clear:  Clear history\n" <>
                          "help:   Display this message\n"

main :: IO ()
main = do
  displayPresent <- lookupEnv "DISPLAY"
  case displayPresent of
    Nothing -> putText "X display not available. Please start Xorg before running greenclip"
    _       -> getArgs >>= run . parseArgs . fmap toS
